bayes_kit/ess.py

```diff
@@ -85,15 +45,15 @@ def ess_ipse(chain: VectorType) -> FloatType:
     Return:
     estimated effective sample size for the specified Markov chain
 
-    Throws:
-    ValueError: if there are fewer than 4 elements in the chain
+    Raises:
+        ValueError: if there are fewer than 4 elements in the chain
     """
     if len(chain) < 4:
-        raise ValueError(f"ess requires len(chains) >=4, but {len(chain) = }")
+        raise ValueError(f"ess requires len(chains) >= 4, but {len(chain)=}")
     acor = autocorr(chain)
-    n = first_neg_pair_start(acor)
-    sigma_sq_hat = acor[0] + 2 * sum(acor[1:n])
-    ess = len(chain) / sigma_sq_hat
+    n = _end_pos_pairs(acor)
+    iat = 2 * acor[0:n].sum() - 1
```

Comment:
Is `iat` a standard notation that we'd expect everybody to recognize? I don't recognize it (which isn't in itself surprising) but it also didn't appear as such when I skimmed the Wikipedia article on sample size determination, I can't find corresponding notation in the Stan reference page on effective sample size, and a google search for "iat statistics" or "iat effective sample size" just pulls up a bunch of stuff about the Implicit Association Test (i.e. psychology), so I think it would be pedagogically helpful to use a more descriptive name (or at least comment it).

(I'm okay with `acor` since it's pretty spelled out by the name and documentation of the function call that populates it.)

Reply:
(Same note probably applies for the imse function below)

Reply:
It's for "integrated autocorrelation time".  Here's a [blog post explaining](https://dfm.io/posts/autocorr/) where @dfm explains.

If "iat" is too short for a local variable name, could you please suggest the shortest thing you'll approve and I'll use that.

We could expose the IAT function itself.  Sometimes people like to report that, too.  Then the definition of ESS simplifies to `ESS = chain_length / IAT`.

Reply:
As with `acorr`, I'm not fussed if there's something more explicit right nearby, it was just as-is this wasn't enough for me to know the right things to google.

I really like the idea of exposing an `integrated_autocorr_time` function--it fixes the naming issue, breaks the math into semantically meaningful components, and also exposes some reusable code that people might want direct access to anyway. Three wins!

Reply:
I'll update and add an IAT function and tests.